### ISSUE TITLE
[Yarp Read] Add support for optional autoconnect

### ISSUE THREAD

**Issue by francesco-romano** [robotology-playground/WBI-Toolbox#83]
We need to extend the behaviour of `yarpRead` block.
Currently the name of the port specified is the port to which the block will connect.
We should link this behaviour (together with the temp port support) only to an `autoconnect` option (checkbox).
In case the user does not set the behaviour to be `autoconnect`, the block should only create the port.
This will ensure the compatibility with the usual yarp modules procedure.

cc @jeljaik

**Comment by jeljaik** [comment]
I'll fix this. I already had to do some pending improvements.

**Comment by iron76** [comment]
:+1: thanks a million @jeljaik

**Comment by jeljaik** [comment]
I have addressed this issue by adding a checkbox called `Autoconnect`. When checked, the user can't specify the destination source, but only the source and size of its output. When unchecked, the user can type the name of the destination port. Upon entering this data, the block will simply create the destination port and it's up to the user to connect the two ports via terminal as `yarp connect [source] [dest]` . In-block documentation has also been updated.

![autoconnect_on](https://cloud.githubusercontent.com/assets/5030039/6897175/34fbd75c-d6f2-11e4-9b35-2eda7e88be52.png)
Screenshot of the updated block with `Autoconnect` on.

![autoconnect_off](https://cloud.githubusercontent.com/assets/5030039/6897182/412883fe-d6f2-11e4-9f25-6500ef5d4162.png)
Screenshot of the updated block with `Autoconnect` off.

Hope this is what you desired :smile: 

Cheers!

**Comment by iron76** [comment]
:+1:

**Comment by jeljaik** [comment]
# Notes
1. When unchecking `Autoconnect` make sure that the now-hidden field `From` is filled with some string, e.g. `'...'`, otherwise you get the error `"Cannot retrieve string from parameter 1!!`. This behavior will be nevertheless improved. 
2. Matlab will crash if you activate the `Autoconnect` option  while specifying the name of a port that you don't want to connect just yet.

**Comment by DanielePucci** [comment]
Hi Jorh,
well done. I think, however, that it could be useful to add an additional output to this block, which informs the user that the port is actually connected. For instance, if the port is disconnected, we receive zeros, but the user does not know if these zeros are actually sent by a sender. Can we add a binary output that acknowledges the connection?

**Comment by francesco-romano** [comment]
@DanielePucci in which of the two cases (or both) you want this output?

**Comment by DanielePucci** [comment]
Both cases? If the autoconnect flag is active, then the binary output can be always true.

**Comment by iron76** [comment]
I think the additional output is just a temporary workaround (an I am not really sure necessary right now). The real solution would be to have a visual output (certain Simulink blocks can change their icon) when data are available on the port.

**Comment by traversaro** [comment]
@iron76 an additional "real" value can be really useful to deal with data still not available on that port (for example: don't start the controller until data is available).

**Comment by traversaro** [comment]
@DanielePucci it is possible that the port never received anything even if you use the "autoconnect" option.

**Comment by iron76** [comment]
@traversaro there is an issue here related to the fact that Simulink always assume that data are available while YARP does not. I think any solution would be a workaround. In any case I don't really like the additional output solution. I am tempted by using NaN.

**Comment by DanielePucci** [comment]
NaN would be fine for me. The main point is to distinguish data sent by someone from fake data. I agree with @traversaro, however, that in some cases it would be useful to have an additional output, which may help develop a diagnostics logic in the controller.

**Comment by jeljaik** [comment]
In "Simulink language", missing data should be represented by NaN. This could definitely be done in conjunction with a change in the color of the block ... although, since the blocks are all colored I'd rather go for a particular icon for this block.

**Comment by jeljaik** [comment]
In this case, as @traversaro mentioned, a connected port that does not stream data would also be clearly visible in the model.

**Comment by francesco-romano** [comment]
to be honest I'm a bit scared about the use of `NaN`s

**Comment by iron76** [comment]
I am more scared about using conventions which might difficult to interpret, such as using an additional binary output. Of course we can document things to allow users retrieving the correct meaning of things.

**Comment by DanielePucci** [comment]
I agree with the general principle. In this specific case, however, I think that it would be more misleading to have NaNs than outputs like "connection" and "readFailure". In the @iron76 direction, we might stream NaNs if either connection = 0, or readFailure=1. In this case, the conventions should be clear enough.

**Comment by francesco-romano** [comment]
@jeljaik your issues here: https://github.com/robotology-playground/WBI-Toolbox/issues/83#issuecomment-87736511 should have been fixed now.
Can you check it?

**Comment by nunoguedelha** [comment]
@francesco-romano  regarding the issue https://github.com/robotology-playground/WBI-Toolbox/issues/83#issuecomment-87736511 =>(1), and the symmetric issue:

* When unchecking Autoconnect make sure that the now-hidden field From is filled with some string, e.g. '...', otherwise you get the error "Cannot retrieve string from parameter 1!!
* When checking Autoconnect make sure that the now-hidden field Dest is filled with some string, e.g. '...', otherwise you get the error "Cannot retrieve string from parameter 2!!

I'm checking if these have been resolved by your fix #90 .

I never reproduced the second issue mentioned by Jorh, even with commit 3123960 from Daniele, before your fix.

**Comment by nunoguedelha** [comment]
I'm still seeing this issue:
* When unchecking Autoconnect make sure that the now-hidden field *From* is filled with some string, e.g. '...', otherwise you get the error "Cannot retrieve string from parameter 1!!

**Comment by francesco-romano** [comment]
I've just test it and it works.. but I was in the `fix/yarpRead` branch. Can you test that you get the same behaviour there? 
Just to understand if it was a merge issue.

**Comment by nunoguedelha** [comment]
Well, I did a git diff between your branch and the master branch and there are no diffs.

**Comment by francesco-romano** [comment]
This is good then :)

But how did you test that?

**Comment by DanielePucci** [comment]
Can we make a simple model-test?

**Comment by nunoguedelha** [comment]
I tested with a very basic model that just uses the default yarpRead block. I checked that the "to" field is filled with the default "..." if it is masked (after I check "Autoconnect" and click OK).
That doesn't happen with the "from" field (if I uncheck "Autoconnect" instead).
Looking at the code `WBCLibrary.mdl` I actually see the line missing (as far as I can understand)...

**Comment by nunoguedelha** [comment]
Don't loose your time with this, I'll make a quick check...

**Comment by nunoguedelha** [comment]
Works fine now. I just pushed the branch "fix/yarpRead_default_from_field" for you to review?

**Comment by francesco-romano** [comment]
@nunoguedelha the `from` field is unused if don't set the `autoconnect` flag. This is way we don't care about it. Further more, leaving the name the user set will allow a faster switch between `autoconnect` values.

To explain better this is the "terminal" behaviour we want to replicate in the `yRead` block:
```bash
1) $ yarp read ... /source/port
2) $yarp read /destination/port
```
So if `autoconnect` is on we want to replicate the first behaviour: we want the user to set an already existent port in the `from` field.
If `autoconnect` is off, we want the user to specify only the destination port. And we don't really care about the `from` port.

Lastly, putting `...` in the from port is wrong. There will never exist a port called `...`.

**Comment by nunoguedelha** [comment]
Yes, you're absolutely right.

So we do agree that the cleanest behaviour of the yard Read bock would be not to check the `from` field or try to create a source port at all when the `autoconnect` flag is set. But currently, it is doing this check, right? Do you get the same error as I do?

**Comment by francesco-romano** [comment]
@nunoguedelha yes ok. I never thought about leaving completely empty the field. 
I think I will edit the C++ anyway. I have to add another feature to the yRead port, so i put you in stand-by :smile:
I will ping you again when I finish adding the feature.

**Comment by nunoguedelha** [comment]
ok thanks :wink:

**Comment by francesco-romano** [comment]
@nunoguedelha  I completely change the yarp read.
I added a new yarp read block. 
I don't know if you want to test it, for us is low priority.

**Comment by nunoguedelha** [comment]
I'll test it this week end if I have some time.

**Comment by francesco-romano** [comment]
Given the new yarp read has been quite stable during the last 6 months of work, I think we can close this issue.
